core/configcenter/configurator.go

```diff
@@ -0,0 +1,144 @@
+package configurator
+
+import (
+	"fmt"
+	"log"
+	"strings"
+	"sync"
+	"sync/atomic"
+
+	"github.com/zeromicro/go-zero/core/conf"
+	"github.com/zeromicro/go-zero/core/configcenter/subscriber"
+	"github.com/zeromicro/go-zero/core/logx"
+	"github.com/zeromicro/go-zero/core/threading"
+)
+
+// Configurator is the interface for configuration center.
+type Configurator[T any] interface {
+	// GetConfig returns the subscription value.
+	GetConfig() (T, error)
+}
+
+type (
+	// Config is the configuration for Configurator.
+	Config struct {
+		// Type is the value type, yaml, json or toml.
+		Type string `json:",default=yaml,options=[yaml,json,toml]"`
+		// Log indicates whether to log the configuration.
+		Log bool `json:",default=ture"`
+	}
+
+	configCenter[T any] struct {
+		conf        Config
+		unmarshaler loaderFn
+
+		subscriber subscriber.Subscriber
+
+		listeners []func()
+		lock      sync.Mutex
+		snapshot  atomic.Value
+	}
+
+	loaderFn func([]byte, any) error
+)
+
+var unmarshalers = map[string]loaderFn{
```

Comment:
maybe we can use registry pattern for unmarshalers, so we can share codecs between configcenter and conf pkg, like viper and grpc https://github.com/spf13/viper/blob/master/internal/encoding/encoder.go#L22

Reply:
got it